### PR TITLE
[Form] deprecate using the date and time types with date objects with not-matching timezones

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | 
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Review by yceruto — APPROVED** [review]

**Comment by xabbuh** [comment]
Status: Needs Review

File: src/Symfony/Component/Form/Extension/Core/Type/TimeType.php [hunk 1]
```diff
@@ -210,6 +210,20 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 new DateTimeToArrayTransformer($options['model_timezone'], $options['model_timezone'], $parts, 'text' === $options['widget'], $options['reference_date'])
             ));
         }
+
+        if (\in_array($options['input'], ['datetime', 'datetime_immutable'], true) && null !== $options['model_timezone']) {
+            $builder->addEventListener(FormEvents::PRE_SET_DATA, static function (FormEvent $event) use ($options): void {
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimeType.php]: This event allows mutating the data, I think we should listen to `POST_SET_DATA` instead.
Unless we have to hook before the transformation process, in such case I would just use a (very) low priority.

File: src/Symfony/Component/Form/Extension/Core/Type/TimeType.php [hunk 2]
```diff
@@ -210,6 +210,20 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                $date = $event->getData();
+
+                if (!$date instanceof \DateTimeInterface) {
+                    return;
+                }
+
+                if ($date->getTimezone()->getName() !== $options['model_timezone']) {
+                    trigger_deprecation('symfony/form', '6.2', sprintf('Using a %s instance with a timezone ("%s") not matching the configured model timezone "%s" is deprecated.', \get_class($date), $date->getTimezone()->getName(), $options['model_timezone']));
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimeType.php]:
```suggestion
                    trigger_deprecation('symfony/form', '6.2', sprintf('Using a "%s" instance with a timezone ("%s") not matching the configured model timezone "%s" is deprecated.', \get_class($date), $date->getTimezone()->getName(), $options['model_timezone']));
```
?
Should it mention that it won't be supported and/or trigger an exception, or being implicit is enough?

**Comment by HeahDude** [comment]
~What about other `Date*` types? Shouldn't this be done at the transformer level so such consistency can be checked in all of them?~

EDIT: no model transformation to hook into for such cases

File: src/Symfony/Component/Form/Extension/Core/Type/TimeType.php (hunk 1 above)
Comment (HeahDude) [reply src/Symfony/Component/Form/Extension/Core/Type/TimeType.php]: Ok, when some initial data is set there are cases no model transformation happens, so we could listen to `POST_SET_DATA` indeed.

**Comment by HeahDude** [comment]
Should we create a dedicated listener that can be used in other `Date*` types as well?

**Comment by xabbuh** [comment]
> Should we create a dedicated listener that can be used in other `Date*` types as well?

I don't think we need to extract this into a dedicated listener, but you are right that we should trigger the deprecation in these types as well. I have updated the code accordingly.

Status: Needs Review

**Review by yceruto — APPROVED** [review]

File: src/Symfony/Component/Form/CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+6.2
+---
+
+ * Deprecate using `DateTime` or `DateTimeImmutable` model data with a different timezone than configured with the
+   `model_timezone` option in `DateType`, `DateTimeType`, and `TimeType`.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/CHANGELOG.md]: no final dot :)
(likely the same in the upgrade file)

**Review by nicolas-grekas — COMMENTED** [review]
What's the motivation for this PR?
What's the alternative way, aka how actionable is this deprecation?
Is this never supposed to happen or is this a common error?

**Comment by xabbuh** [comment]
> What's the motivation for this PR?
> What's the alternative way, aka how actionable is this deprecation?
> Is this never supposed to happen or is this a common error?

I discovered this while working on some form related tests recently where tests failed on GHA while passing locally. From my experience such failures are hard to track down if you have them on production servers but not locally. Now you would at least see more information in the deprecation log that make you aware of a potential issue.

What I can imagine as an alternative is being able to omit the `model_timezone` option when the input is a date object and let the timezone default to the timezone of the date object.

**Comment by nicolas-grekas** [comment]
> What I can imagine as an alternative is being able to omit the model_timezone option when the input is a date object and let the timezone default to the timezone of the date object.

Does this mean that the `model_timezone` option shouldn't exist at all, if the datetime object already embeds it?

If yes, let's deprecate it?

If no, what's its purpose? And can you add the new envisioned behavior as commented code? It's going to be a `throw` isn't it?

**Comment by stof** [comment]
Well, model_timezone is still very useful for the reverse transformation, to know which timezone should be used as the target of the transformation (this one cannot be inferred from the DateTime object, as we need to produce it)

**Comment by xabbuh** [comment]
We also need it for any other input than `DateTime`/`DateTimeImmutable` (i.e. `string` and `array`).

I am going to update the PR later to address your feedback, Nicolas.

**Comment by nicolas-grekas** [comment]
Friendly ping @xabbuh

**Comment by fabpot** [comment]
Is this one ready for another round of reviews?

**Comment by xabbuh** [comment]
Yes, the PR is ready to be reviewed again.

File: UPGRADE-6.3.md [hunk 4]
```diff
@@ -7,6 +7,12 @@ DependencyInjection
  * Deprecate `PhpDumper` options `inline_factories_parameter` and `inline_class_loader_parameter`, use `inline_factories` and `inline_class_loader` instead
  * Deprecate undefined and numeric keys with `service_locator` config, use string aliases instead
 
+Form
+----
+
+* Deprecate using `DateTime` or `DateTimeImmutable` model data with a different timezone than configured with the
```
Comment (nicolas-grekas) [inline UPGRADE-6.3.md]: missing space at start of line

File: src/Symfony/Component/Form/Extension/Core/Type/DateTimeType.php [hunk 5]
```diff
@@ -195,6 +197,20 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+        if (\in_array($options['input'], ['datetime', 'datetime_immutable'], true) && null !== $options['model_timezone']) {
+            $builder->addEventListener(FormEvents::POST_SET_DATA, static function (FormEvent $event) use ($options): void {
+                $date = $event->getData();
+
+                if (!$date instanceof \DateTimeInterface) {
+                    return;
+                }
+
+                if ($date->getTimezone()->getName() !== $options['model_timezone']) {
+                    trigger_deprecation('symfony/form', '6.3', sprintf('Using a "%s" instance with a timezone ("%s") not matching the configured model timezone "%s" is deprecated.', $date::class, $date->getTimezone()->getName(), $options['model_timezone']));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/Type/DateTimeType.php]: I think we're missing some inline comment to know what to do with this in 7.0.
Are we going to throw something? If yes, can we add the anticipated line as a comment already?
Same for tests: how should they be changed in 7.0? Let's add some commented expectException?

**Comment by nicolas-grekas** [comment]
friendly ping @xabbuh  :)

File: src/Symfony/Component/Form/Extension/Core/Type/DateTimeType.php (hunk 5 above)
Comment (xabbuh) [reply src/Symfony/Component/Form/Extension/Core/Type/DateTimeType.php]: done

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @xabbuh.

**Comment by fmarchalemisys** [comment]
Hi guys,

This change is breaking our app because `DateTime` comes from various sources (php, doctrine, external modules). The timezone can either be utc or the server timezone depending on what created the php `DateTime`.

Having to normalize the timezone of a hundred of `DateTime` before passing them to nearly as many `FormType` would requires a substantial effort.

Can you elaborate on the issues that prompted this PR? In short, is throwing an exception to the face of the user the right approach?

**Comment by xabbuh** [comment]
@fmarchalemisys Please open a new issue. Comments on merged PRs are likely to get lost.

**Pull request merged**